unknown regression in 0.5.7 release

After upgrading to the `0.5.7` series, I can no longer compile [michaelficarra/purescript-spidermonkey-ast](https://github.com/michaelficarra/purescript-spidermonkey-ast) due to what I believe is a bug in the type checker. The error I get is

```
[...]
Error checking type of term Prelude.($)(SpiderMonkeyAST.toJSON)(SpiderMonkeyAST.unread(a)) against type Prim.String
Error in expression Prelude.($)(SpiderMonkeyAST.toJSON)(SpiderMonkeyAST.unread(a)):
Error checking type of term Prelude.($)(SpiderMonkeyAST.toJSON)(SpiderMonkeyAST.unread(a)) against type Prim.String
Error in expression Prelude.($)(SpiderMonkeyAST.toJSON):
Error applying function of type u8102 -> u8101 to argument SpiderMonkeyAST.unread(a)
Error in expression SpiderMonkeyAST.unread(a):
Error checking type of term SpiderMonkeyAST.unread(a) against type u8104
Error in expression SpiderMonkeyAST.unread(a):
Error inferring type of value
Error in expression SpiderMonkeyAST.unread:
Error applying function of type SpiderMonkeyAST.ObjectProperty -> SpiderMonkeyAST.SMAST to argument a
Error in expression a:
Error checking type of term a against type SpiderMonkeyAST.ObjectProperty
Error in expression a:
Error checking that type SpiderMonkeyAST.Node subsumes type SpiderMonkeyAST.ObjectProperty
Error unifying type SpiderMonkeyAST.Node with type SpiderMonkeyAST.ObjectProperty
Cannot unify SpiderMonkeyAST.Node with SpiderMonkeyAST.ObjectProperty.
make: *** [lib/SpiderMonkeyAST.js] Error 1
```

The key line is `Error checking type of term a against type SpiderMonkeyAST.ObjectProperty` in the `Show` instance. I've reduced the program as much as I can to

```
module SpiderMonkeyAST where

import Data.Array () -- for Functor instance of Array

data P = Q { n :: N }
data N = O { ps :: [P] }

foreign import data S :: *

foreign import pToS' :: { n :: S } -> S
pToS (Q p) = pToS' { n: nToS p.n }

foreign import nToS' :: forall a. { x :: String | a } -> S
nToS :: N -> S
nToS (O a) = nToS' { x: "O", ps: pToS <$> a.ps }

foreign import sToString :: S -> String
instance showN :: Show N where
  show a = sToString $ nToS a
```

**edit:** updated the example to be as minimal as possible

--- a/examples/passing/MutRec2.purs
+++ b/examples/passing/MutRec2.purs
@@ -1,0 +1,17 @@
+module Main where
+
+data A = A B
+ 
+data B = B A
+
+foreign import data S :: *
+
+f :: A -> S
+f a = case a of A b -> g b
+
+g b = case b of B a -> f a
+
+showN :: A -> S
+showN a = f a
+
+main = Debug.Trace.trace "Done"

--- a/examples/passing/MutRec3.purs
+++ b/examples/passing/MutRec3.purs
@@ -1,0 +1,17 @@
+module Main where
+
+data A = A B
+ 
+data B = B A
+
+foreign import data S :: *
+
+f a = case a of A b -> g b
+
+g :: B -> S
+g b = case b of B a -> f a
+
+showN :: A -> S
+showN a = f a
+
+main = Debug.Trace.trace "Done"

--- a/src/Language/PureScript/TypeChecker.hs
+++ b/src/Language/PureScript/TypeChecker.hs
@@ -183,7 +183,11 @@ typeCheckAll mainModuleName moduleName exps = go
       forM_ (map (\(ident, _, _) -> ident) vals) $ \name ->
         valueIsNotDefined moduleName name
       tys <- typesOf mainModuleName moduleName $ map (\(ident, _, ty) -> (ident, ty)) vals
-      vals' <- forM (zipWith (\(name, nameKind, _) (_, (val, ty)) -> (name, val, nameKind, ty)) vals tys) $ \(name, val, nameKind, ty) -> do
+      vals' <- forM [ (name, val, nameKind, ty) 
+                    | (name, nameKind, _) <- vals
+                    , (name', (val, ty)) <- tys
+                    , name == name' 
+                    ] $ \(name, val, nameKind, ty) -> do
         addValue moduleName name ty nameKind
         return (name, nameKind, val)
       return $ BindingGroupDeclaration vals'

